src/bin/pgcopydb/ld_store.c

```diff
@@ -0,0 +1,446 @@
+/*
+ * src/bin/pgcopydb/ld_store.c
+ *     Implementation of a CLI to copy a database between two Postgres instances
+ */
+
+#include <errno.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "parson.h"
+
+#include "copydb.h"
+#include "ld_stream.h"
+#include "lock_utils.h"
+#include "log.h"
+#include "parsing_utils.h"
+#include "pg_utils.h"
+#include "schema.h"
+#include "signals.h"
+#include "string_utils.h"
+
+
+/*
+ * ld_store_open_replaydb opens the current replaydb file if it already exists,
+ * or create a new replaydb SQLite file for processing the streaming data.
+ */
+bool
+ld_store_open_replaydb(StreamSpecs *specs)
+{
+	StreamContext *privateContext = &(specs->private);
+	DatabaseCatalog *replayDB = specs->replayDB;
+
+	if (!ld_store_current_filename(specs))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	bool createReplayDB = IS_EMPTY_STRING_BUFFER(replayDB->dbfile);
+
+	/* if we don't have a replayDB filename yet, it's time to create it */
+	if (IS_EMPTY_STRING_BUFFER(replayDB->dbfile))
+	{
+		if (privateContext->timeline == 0)
+		{
+			log_error("BUG: ld_store_open_replaydb: timeline is zero");
+			return false;
+		}
+
+		sformat(replayDB->dbfile, MAXPGPATH, "%s/%08d-%08X-%08X.db",
+				specs->paths.dir,
+				privateContext->timeline,
+				LSN_FORMAT_ARGS(privateContext->startpos));
+	}
+
+	log_info("%s CDC file \"%s\"",
+			 createReplayDB ? "Creating" : "Opening",
+			 replayDB->dbfile);
+
+	/* now open the replaydb */
+	if (!catalog_init(replayDB))
+	{
+		log_error("Failed to open the current replay database \"%s\", "
+				  "see above for details",
+				  replayDB->dbfile);
+		return false;
+	}
+
+	if (createReplayDB)
+	{
+		if (!ld_store_insert_cdc_filename(specs))
+		{
+			log_error("Failed to register the current replay database \"%s\", "
+					  "see above for details",
+					  replayDB->dbfile);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * ld_store_current_filename queries the sourceDB SQLite catalog for an open
+ * file for the current timeline.
+ */
+bool
+ld_store_current_filename(StreamSpecs *specs)
+{
+	StreamContext *privateContext = &(specs->private);
+
+	sqlite3 *db = specs->sourceDB->db;
+
+	if (db == NULL)
+	{
+		log_error("BUG: ld_store_current_filename: db is NULL");
+		return false;
+	}
+
+	char *sql =
+		"  select filename "
+		"    from cdc_files "
+		"   where done_time_epoch is null "
+		"     and startpos <= $1 "
+		"     and (endpos is null or $2 <= endpos) "
```

Comment:
@dimitri I'm wondering how comparison works here? AFAIK, SQlite stores pg_lsn type as just a blob and we don't seem to provide comparison operator either. Probably because Sqlite  uses `memcmp` to compare blobs and it works correctly on hex string?

Reply:
That's the gist of it yeah, LSN as text in hexadecimal still allows using the usual operators to compare values... (I believe?)